text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+	($x:tt) => {}
+}
+single_tt!(enum#discrimant);
+```
+now matches, instead of being a lexical error.
+
+In editions 2015 and 2018, this feature is not available.
+
+## Parsing
+
+A new form of expression is added,
+
+> *DiscriminantExpression* :
+> > *Expression* `.` `enum#discriminant`
+
+Like `.await`, this is *not* a place expression, and as such is invalid on the
+left-hand side of an assignment, giving an error like the following:
+
+```text
+error[E0070]: invalid left-hand side of assignment
+ --> src/lib.rs:5:29
+  |
+5 |         x.enum#discriminant = 4;
+  |         ------------------- ^
+  |         |
+  |         cannot assign to this expression
+```
+
+## Visibility
+
+This acts as though it were a `pub(in self)` field on a type.
+
+As such, it's an error to use `.enum#discriminant` on types from sub-modules or other crates.
+
+```rust
+mod inner {
+	pub enum Foo { Bar }
+}
+inner::Foo::Bar.enum#discriminant // ERROR: enum discriminant is private
+
+```
+
+## Type
+
+The LHS is auto-deref'd until it finds something known to be an `enum`.
+
+*Note: this is different from `mem::discriminant`.  For example,*
+```rust
+#![allow(enum_intrinsics_non_enums)]
+enum MyEnum { A, B }
+let a = Box::new(MyEnum::A);
+let b = Box::new(MyEnum::B);
+assert_eq!(std::mem::discriminant(&a), std::mem::discriminant(&b));
+assert_ne!(a.enum#discriminant, b.enum#discriminant);
+```
+<!-- https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=76293c2f83cb7719c22c15bcebeaeb13 -->
+
+For this, a generic parameter is never considered to be an `enum`,
+although a generic enum where some of the generic parameters to the
+enum constructor are not yet known is fine.
+
+It's an error if, despite deref'ing, the LHS is still not an `enum`.
+
+If the enum has `repr(uN)` or `repr(iM)`, the `.enum#discriminant` expression
+returns a value of type `uN` or `iM` respectively.
+
+If the enum does not specify an integer `repr`, then it returns `isize`.
+
+*Note: `isize` is rarely the desired type for discriminants, and indeed custom
+discriminants on types with fields are disallowed without explicit `repr` types.
+Returning `isize` is fine here, though, thanks to privacy because the code
+inside the module can be updated should it change to specify a specific type.*
+
+## Semantics
+
+When the LHS of a discriminant expression is a *place*, that place is read but not consumed.
+
+*Note: this can be thought of as if it read a field of `Copy` type from the LHS.*
+
+This lowers to [`Rvalue::Discriminant`][MIR discr] in MIR.
+
+[MIR discr]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html#variant.Discriminant
+
+As this expression is an *r-value*, not a *place*, `&foo.enum#discriminant` returns
+a reference to a temporary, aka is the same as `&{foo.enum#discriminant}`.
+It does *not* return a reference to the memory in which the discriminant is
+stored -- not even for types that do store the discriminant directly.
```

Comment:
Instead of making this an rvalue, I think we should make it a "read-only move-only place" -- a place you can read from but not write to or create a pointer to. That would be more forward-compatible with actually making this a move-only place once these exist. The difference is that with a move-only place I would think `&foo.enum#discriminant` is an error rather than creating a temporary.

`let m = &mut foo.enum#discriminant; *m = 5;` would be accepted under your proposal, but probably not behave the way the user expects. If they *want* the copy they can always write `let m = &mut {foo.enum#discriminant};` (and the error message can indicate that).

Reply:
What is the advantage of a "read-only move-only place" over a (r-)value expression?

Reply:
I've spelled this out, I think?

1. let m = &mut foo.enum#discriminant; *m = 5; would be accepted under your proposal, but probably not behave the way the user expects. 
2. That would be more forward-compatible with actually making this a move-only place once these exist.

Reply:
I think a benefit of special method `.enum#discriminant()` over a special field `.enum#discriminant` is that the reference footgun noted above is avoided without introducing any new features. Users already know that method calls produce temporary values.